### PR TITLE
Create REST Service

### LINKED ISSUE #1: Create REST Server

### PR DESCRIPTION
**Pull request by ric-evans**
closes #1

### CODE REVIEW

**Comment by ric-evans** [comment]
**TODO:** integration tests

**Comment by ric-evans** [comment]
@dsschult I'm going to add more failure tests, but it's in a stable place now

File: rest_server/handlers.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    async def post(self) -> None:
+        """Start a new scan."""
+
+        def valid_event_id(val: Any) -> str:
+            val = str(val)
+            if any(x.isspace() for x in val) or not any(x.isalnum() for x in val):
+                raise TypeError(
+                    "event_id must contain alphanumeric characters and no spaces"
+                )
+            return val
```
Comment (dsschult) [inline rest_server/handlers.py:113]: Couldn't this whole function be replaced by `str(val).isalnum()`?  That already checks for spaces and can operate on a whole string.

File: .github/workflows/wipac-cicd.yml [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        py3: ${{ fromJSON(needs.py-versions.outputs.matrix) }}
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v3
+        with:
+          python-version: ${{ matrix.py3 }}
+
+      - name: Start MongoDB
+        uses: supercharge/mongodb-github-action@1.8.0
+        # TODO: change port/username/password?
```
Comment (dsschult) [inline .github/workflows/wipac-cicd.yml:100]: This should be a service container. Example:
https://github.com/Observation-Management-Service/ewms-queue-benchmarking/blob/main/.github/workflows/wipac_cicd.yaml#L55

File: rest_server/database/interface.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+        if not progress:
+            msg = f"Attempted progress update with an empty object ({progress})"
+            raise web.HTTPError(
+                422,
+                log_message=msg + f" for {scan_id=}",
+                reason=msg,
+            )
+        manifest = await self.find_one(scan_id, True)
+        manifest.progress = progress
+        manifest = await self.upsert(manifest)
```
Comment (dsschult) [inline rest_server/database/interface.py:143]: This will suffer from race conditions with the two independent DB calls.

File: rest_server/database/interface.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        return manifest
+
+    async def mark_as_deleted(self, scan_id: str) -> schema.Manifest:
+        """Mark `schema.Manifest` at doc matching `scan_id` as deleted."""
+        LOGGER.debug(f"marking manifest as deleted for {scan_id=}")
+        manifest = await self.find_one(scan_id, True)
+        manifest.is_deleted = True
+        manifest = await self.upsert(manifest)
```
Comment (dsschult) [inline rest_server/database/interface.py:151]: Same race condition problem.

File: rest_server/database/interface.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        result = schema.Result(scan_id, False, json_dict)
+        result = await self.upsert(result)
+        return result
+
+    async def mark_as_deleted(self, scan_id: str) -> schema.Result:
+        """Mark `schema.Result` at doc matching `scan_id` as deleted."""
+        LOGGER.debug(f"marking result as deleted for {scan_id=}")
+        result = await self.find_one(scan_id, True)
+        result.is_deleted = True
+        result = await self.upsert(result)
```
Comment (dsschult) [inline rest_server/database/interface.py:210]: More race conditions.

File: rest_server/server.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    args["mongodb_url"] = mongodb_url()
+
+    # Configure REST Routes
+    rs = RestServer(debug=debug)
+    for name, klass in inspect.getmembers(
+        handlers,
+        predicate=lambda x: (
+            inspect.isclass(x) and issubclass(x, RestHandler) and x != RestHandler
+        ),
+    ):
```
Comment (dsschult) [inline rest_server/server.py:53]: One thing you may or may not run into with this fancy auto-detection of routes is if ordering becomes important.  If one of the routes is more of a catch-all in the regular expression, it could hide other routes if it goes before them in getting added.

**Review by dsschult — COMMENTED** [review]
One thing I'm not sold on is the multiple levels of indirection before you get to the mongo calls.  This seems both inefficient and hides what is really going on (somewhat hard to follow, at least to me).  And it can lead to situations where race conditions can develop.

File: rest_server/handlers.py (hunk 1 above)
Comment (ric-evans) [reply rest_server/handlers.py:113]: I replaced the error message with "must contain no spaces and at least one alphanumeric character"

Comment (ric-evans) [reply rest_server/handlers.py:113]: I figure event ids like `abc.2` and `foobar[baz]` are okay

File: rest_server/server.py (hunk 6 above)
Comment (ric-evans) [reply rest_server/server.py:53]: That's a really good point, I hadn't thought of that

File: rest_server/handlers.py (hunk 1 above)
Comment (dsschult) [reply rest_server/handlers.py:113]: Maybe.  The check is just really vague right now.  Would be nice to get this tightened up somehow.

Comment (ric-evans) [reply rest_server/handlers.py:113]: I'm not sure what we want to allow in an id. For the actual logic, preferably this would be a `re.match`

File: rest_server/database/interface.py (hunk 3 above)
Comment (ric-evans) [reply rest_server/database/interface.py]: Fixed :heavy_check_mark:

File: rest_server/database/interface.py (hunk 4 above)
Comment (ric-evans) [reply rest_server/database/interface.py]: Fixed :heavy_check_mark:

File: rest_server/database/interface.py (hunk 5 above)
Comment (ric-evans) [reply rest_server/database/interface.py]: Fixed :heavy_check_mark:

**Comment by ric-evans** [comment]
> One thing I'm not sold on is the multiple levels of indirection before you get to the mongo calls. This seems both inefficient and hides what is really going on (somewhat hard to follow, at least to me). And it can lead to situations where race conditions can develop.

I removed one level, I think it clears up the most confusing part

File: rest_server/__main__.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        await asyncio.Event().wait()
+    finally:
+        await rs.stop()  # type: ignore[no-untyped-call]
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-l", "--log", default="DEBUG", help="the output logging level")
+    _args = parser.parse_args()
```
Comment (dsschult) [inline rest_server/__main__.py:24]: Is the only way to configure logging via the cmd line arg?  This seems odd since we take everything else via env variable.

File: rest_server/config.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+EXCLUDE_DBS = [
+    "system.indexes",
+    "production",
+    "local",
+    "config",
+    "token_service",
+    "admin",
+]
+
+EXCLUDE_COLLECTIONS = ["system.indexes"]
```
Comment (dsschult) [inline rest_server/config.py:46]: What are these two constants used for?  They don't seem to show up in the code.

File: rest_server/config.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    MONGODB_AUTH_PASS: str = ""  # empty means no authentication required
+    MONGODB_AUTH_USER: str = ""  # None means required to specify
+    MONGODB_HOST: str = "localhost"
+    MONGODB_PORT: int = 27017
+    REST_HOST: str = "localhost"
+    REST_PORT: int = 8080
+    CI_TEST: bool = False
+
+
+ENV = from_environment_as_dataclass(EnvConfig)
```
Comment (dsschult) [inline rest_server/config.py:31]: While this pattern of a global env variable is mostly ok, one thing it can't do is changing them during testing.  Something to be aware of if you end up needing to do so down the line.

File: rest_server/handlers.py (hunk 1 above)
Comment (dsschult) [reply rest_server/handlers.py:105]: Is this supposed to be an i3 event id, or something we make up?

File: rest_server/__main__.py (hunk 7 above)
Comment (ric-evans) [reply rest_server/__main__.py:24]: I made it an env variable

File: rest_server/config.py (hunk 8 above)
Comment (ric-evans) [reply rest_server/config.py:46]: That's copypasta from the mou dash

**Comment by ric-evans** [comment]
https://github.com/WIPACrepo/SkyDriver/pull/2#discussion_r1048713480

File: rest_server/config.py (hunk 9 above)
Comment (ric-evans) [reply rest_server/config.py:35]: Yeah, that's a good point--a definite downside for testing. The alternative is to make `ENV` a function and re-fetch the env vars, but I think that leads to more problems in the production code since the config state is now mutable.

**Comment by dsschult** [comment]
> [#2 (comment)](https://github.com/WIPACrepo/SkyDriver/pull/2#discussion_r1048713480)
> 
> It's the i3 event id or the realtime event id. The idea is to keep it pegged to something external, so users can query scans using it.

I guess that would be `<int>.<int>`?  Though if we really don't care what it is, maybe use it as an opaque string?

File: rest_server/config.py (hunk 9 above)
Comment (dsschult) [reply rest_server/config.py:35]: The other thing I've done is only get the env vars when needed (usually when making the server).  That technically means it's mutable if you restart the server, but you only do that in testing.
